fix(transformer/arrow-functions): `_this = this` should be inserted after super call expression

Dunqing opened the pull request:
related: #7792 

This PR doesn't contain fixing the async arrow function in `super()`, which is difficult for our architecture. I just found that `esbuild`'s implementation is quite simpler! https://esbuild.github.io/try/#dAAwLjI0LjAALS10YXJnZXQ9ZXMyMDE2AGNsYXNzIEMgZXh0ZW5kcyBTIHsKICBjb25zdHJ1Y3RvcigpIHsKICAgIHN1cGVyKGFzeW5jICgpID0+IHRoaXMpOwogICAgYXN5bmMoKSA9PiB7fQogIH0KfQ

Dunqing commented:
* **#8024** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8024?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8024?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8024).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented on crates/oxc_transformer/src/common/arrow_function_converter.rs:
```diff
@@ -703,6 +728,25 @@ impl<'a> ArrowFunctionConverter<'a> {
+            if call.callee.is_super() {
+                let scope_id = ctx.current_scope_id();
+                self.super_call_addresses.entry(scope_id).or_insert_with(|| {
+                    ctx.ancestors()
+                        .find(|ancestor| {
+                            matches!(
+                                ancestor,
+                                // const A = super():
+                                Ancestor::VariableDeclarationDeclarations(_)
+                                // super();
+                                | Ancestor::ExpressionStatementExpression(_)
+                            )
+                        })
+                        .unwrap()
+                        .address()
```
Unwrap failed here, we need to find all possible statements, and this may not be a good implementation

overlookmotel replied:
You can use `Ancestor::is_via_statement` for this.

Note it doesn't tell you if that ancestor *is* a statement, but whether its *child* is a statement.

So if you are visiting a `Function`, it could be either an expression or a statement. `ctx.parent().is_via_statement()` will tell you.

To find the closest statement, loop through ancestors, and when `ancestor.is_via_statement()` returns `true`, then the *previous* ancestor you visited is the statement.

The naming `is_via_*` is weird, and the fact that you have to look at the ancestor above is confusing as hell. So it's a terrible API! But it does do what you need.

overlookmotel replied:
#8031 renames this method to `Ancestor::is_parent_of_statement`. I think that's clearer.

Dunqing replied:
I just found that we have these methods!

Dunqing commented:
@overlookmotel Can you take a look at this PR, I am not sure this is a good. I really want to change everything according to esbuild's implementation

overlookmotel commented:
Concerning ESBuild's version:

Yes it's simpler to implement, but I think less performant at runtime than Babel's version. In Babel's version, there's only 1 call to `asyncToGenerator` when the async function is defined. But in ESBuild's version, `__async` gets called and constructs a new generator function *every time you call the async function*.

I don't know for sure, but that sounds more expensive (more garbage collection).

I think we could do our own version which is better than both! Maybe something like this:

Input:

```js
class C extends S {
  constructor() {
    super();
    let f = async () => this;
  }
}
```

Output:

```js
class C extends S {
  constructor() {
    var _getThis = () => this;
    super();
    let f = oxcHelpers.asyncToGenerator(function*() { return this; }, _getThis, "f");
  }
}
```

(the last param takes care of naming the function correctly, instead of Babel's extra function wrappers)

But... I suggest we follow Babel for now, make our implementation correct, and look at optimizing output later. There's many places we can improve on Babel in *lots* of transforms.

overlookmotel commented:
PS: I've made a small improvement to the `super()`-replacer visitor in #8028.

Dunqing commented:
That's a trade-off for us considering the different implementations. When I am a developer for the transformer, I prefer a 
way easy to implement save my hair!

Dunqing commented:
To clarify, so in this case, you prefer to double-visited? I have no strong opinion about this. Just think of this way is easier to implement, but cannot solve `super(async () => {})`.

overlookmotel commented:
I think double-visit is the best way, yes.

I *think* it'll be able to handle the `super(async () => this)` case too.

We could later try and optimize it for common case where `super()` is a top level statement, by storing the `Address` of `super()` during traversal, and then you know exactly where to insert `_this = this` without another visit.

But:

1. I'm not actually sure if that'll be a perf gain or not, because you have to do a lot more checking for "is this node `super()`?", when most of the time the answer is "no". And when the class constructor *doesn't* contain an async arrow function (most won't), then that's wasted work.
2. That won't handle all the edge cases anyway.
3. Make it correct first, then optimize.

Dunqing commented:
I have reimplemented the fixer in https://github.com/oxc-project/oxc/pull/8024/commits/89eb36f541d6ca4381e827580834eceabd01e1f1. The new implementation covers `super(async () => { this })` case

overlookmotel reviewed (approved):
Very nice!

I pushed a couple of commits to extend the tests, and will follow-up with some refactoring PRs. But it's all small stuff. Overall, looks great.

overlookmotel commented:
### Merge activity

* **Jan 7, 8:45 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
fix(transformer/arrow-functions): `_this = this` should be inserted after super call expression (#8024)

related: #7792

This PR doesn't contain fixing the async arrow function in `super()`, which is difficult for our architecture. I just found that `esbuild`'s implementation is quite simpler! https://esbuild.github.io/try/#dAAwLjI0LjAALS10YXJnZXQ9ZXMyMDE2AGNsYXNzIEMgZXh0ZW5kcyBTIHsKICBjb25zdHJ1Y3RvcigpIHsKICAgIHN1cGVyKGFzeW5jICgpID0+IHRoaXMpOwogICAgYXN5bmMoKSA9PiB7fQogIH0KfQ
